src/router.rs

```diff
@@ -1,127 +1,63 @@
-use crate::{alcazar::AlcazarError, http_request::HttpError};
+use crate::error::{AlcazarError, HttpError, Result};
+use crate::routing::endpoint::{Endpoint, MethodType};
 use std::str::FromStr;
 
-#[derive(PartialEq, Copy, Clone)]
-pub enum MethodType {
-    POST,
-    GET,
-    PATCH,
-    DELETE,
-    CONNECT,
-    OPTIONS,
-    TRACE,
-    HEAD,
-}
-
-impl FromStr for MethodType {
-    type Err = AlcazarError;
-
-    fn from_str(method: &str) -> Result<MethodType, AlcazarError> {
-        match method {
-            "POST" => Ok(MethodType::POST),
-            "GET" => Ok(MethodType::GET),
-            "PATCH" => Ok(MethodType::PATCH),
-            "DELETE" => Ok(MethodType::DELETE),
-            "CONNECT" => Ok(MethodType::CONNECT),
-            "OPTIONS" => Ok(MethodType::OPTIONS),
-            "TRACE" => Ok(MethodType::TRACE),
-            "HEAD" => Ok(MethodType::HEAD),
-            _ => Err(AlcazarError::HttpError(HttpError::MethodNotImplemented)),
-        }
-    }
-}
-
 #[derive(Clone)]
-pub struct Endpoint {
-    method: MethodType,
-}
-
-impl Default for Endpoint {
-    fn default() -> Self {
-        Self {
-            method: MethodType::GET,
-        }
-    }
-}
-
-impl Endpoint {
-    pub fn new() -> Self {
-        Endpoint::default()
-    }
-
-    pub fn set_method(mut self, method: MethodType) -> Self {
-        self.method = method;
-        self
-    }
-}
-
-#[derive(Clone)]
-pub struct Route {
-    path: String,
-    endpoint: Endpoint,
+pub struct Router {
+    endpoints: Vec<Endpoint>,
 }
 
-impl Default for Route {
+impl Default for Router {
     fn default() -> Self {
         Self {
-            path: "/".into(),
-            endpoint: Endpoint::default(),
+            endpoints: Vec::new(),
         }
     }
 }
 
-impl Route {
+impl Router {
     pub fn new() -> Self {
-        Route::default()
-    }
-
-    pub fn set_path(mut self, path: String) -> Self {
-        self.path = path;
-        self
-    }
-
-    pub fn set_endpoint(mut self, endpoint: Endpoint) -> Self {
-        self.endpoint = endpoint;
-        self
-    }
-
-    pub fn get_response(&self) -> &'static str {
-        "HTTP/1.1 200 OK\r\n\r\n"
+        Router::default()
     }
-}
-
-#[derive(Clone)]
-pub struct Router {
-    routes: Vec<Route>,
-}
 
-impl Default for Router {
-    fn default() -> Self {
-        Self { routes: Vec::new() }
+    // Returns a list of declared endpoints.
+    pub(crate) fn endpoints(&self) -> &Vec<Endpoint> {
+        &self.endpoints
     }
-}
 
-impl Router {
-    pub fn new() -> Self {
-        Router::default()
-    }
+    // TODO: Add handler parameter and set it up
+    pub fn with_endpoint(mut self, path: &str, methods: &[&str]) -> Self {
```

Comment:
The methods argument could be a `AsRef<[&str]>`, because then users can pass a array to the method.

Reply:
Actually this can be `AsRef<str>` I don't think this method takes multiple endpoint, but yes passing AsRef would be nice.

Reply:
Specifiying an argument type as a construction like `AsRef<[&str]>` will lead to a compilation error, because compiler doesn't know the size for values there (the must be wrapped in `Box<T>` or something to solve it).
Therefore, it will be much easier to everyone to keep it simple as much as possible. In addition to it, this method will be used only for declaring endpoint in place of the Router declaration.